### ISSUE TITLE
mof_compiler; need to standardize the cmd line options to match our std

### ISSUE THREAD

**Issue by KSchopmeyer** [pywbem/pywbem#216]
Document what we want, and whether it should be in the url, etc.

**Comment by KSchopmeyer** [comment]
In discussion on 13 April telecon we concluded that the standard within pywbem should be to try to use the uri for everything and avoid extra parameters as much as possible.  Therefore, we need to provide alternatives for those script calls that use things like --no-ssl so that the url itself can be used.

**Comment by KSchopmeyer** [comment]
Right now, I believe that this is wbemcli at least that uses the --no-ssl.

Somewhere we probably need to start our own style guide to capture style decisions like this so that they are general knowledge for the future.

**Comment by andy-maier** [comment]
**Proposal:**
For specifying WBEM server & connection related things, I suggest to use this:

    -s, --server server
            URL of the WBEM server.
            - Scheme is optional and defaults to HTTP.
            - Host is required.
            - Port is optional and defaults to the standard ports, depending on HTTP/HTTPS.
    -n, --namespace namespace
            Namespace in the WBEM server to work against
            Default: root/cimv2
    -u, --username username
            Username for authenticating with the WBEM server
            Default: No username
    -p, --password password
            Password for authenticating with the WBEM server
            Default: Prompt for password if a username was specified
    -c, --cert certfile
            File path of certificate file to be used with HTTPS.
    -k, --key keyfile
            File path of key file to be used with HTTPS.
    -t, --timeout timeout
            Timeout in seconds
            Default: TBD

**Discussion:** Should we change wbemcli and mof_compiler to support only this proposal?

**Comment by andy-maier** [comment]
Some data points:
* The current [PyWBEM utilities](https://pywbem.readthedocs.org/en/latest/utilities.html)
* [OpenPegasus cimcli man page](http://cvs.opengroup.org/cgi-bin/viewcvs.cgi/pegasus/src/Clients/cimcli/doc/Attic/cimcli.html?hidecvsroot=1&search=None&hideattic=1&sortby=file&logsort=date&rev=1.1&diff_format=h)
* [SFCB wbemcli](manpages.ubuntu.com/manpages/precise/man1/wbemcli.1.html)

**Comment by andy-maier** [comment]
On compatibility:

* mof_compiler could find out that it is being invoked with the old options if:
  - the parameter after `-u`has URI format, or
  - the old option `--url` is specified, or
  - the required `-s` or `--server` option is missing
* wbemcli could find out that it is being invoked with the old options if:
  - one of the old options `--port` or `--no-ssl` are used, or
  - the required `-s` or `--server` option is missing.

In those cases, the command could error out indicating old options used.

**Comment by andy-maier** [comment]
Also: We could support environment variables that provide defaults for these options:

* PYWBEM_SERVER=server
* PYWBEM_NAMESPACE=namespace
* PYWBEM_USERNAME=username
* PYWBEM_PASSWORD=password
* PYWBEM_CERTFILE=certfile
* PYWBEM_KEYFILE=keyfile
* PYWBEM_TIMEOUT=timeout

**Comment by andy-maier** [comment]
Yet another option would be to have a config file with defaults.

The `configparse` package supports config files, and there is also a package that integrates config file processing and argparse.

**Comment by andy-maier** [comment]
**Discussion:**

I think, first we need to decide upon the externals, leaving things like which package do we use, out for the moment:

* Which set of command line options (see proposal above)?
* How should we deal with old options?
  - Ignore
  - Detect and reject
  - Full backwards compatibility (hard to achieve, given that we change the meaning of some options)
* Should we support config files?
* Should we support env vars?

**Comment by andy-maier** [comment]
I grabbed this one, hope this is ok for you Karl.

**Comment by KSchopmeyer** [comment]
1. Need to add runoperations... to cli list.
2. Pegasus wrote document of cli standards.  Karl will provide.
3. We agreed where possible to keep old options but mark them deprecated.  There will be cases where we have incompatibilities and cannot keep the old options however.
4. We do not understand why sfcb uses a CA cert file.
5. We need to consider the option of cert repository rather than individual cert files.
6. need to determine where we stand and what we want to do about client side basic and digest auth
7. Need to consider whether we should add the ability to get from a config file/ env variables as part of the options. Note Karl found in addition to config package another package that merges argparse and config into a single integrated package. 

Karl will make proposal next week.

**Comment by KSchopmeyer** [comment]
I would like to break this into multiple issues

1. Common base for all cmd line args with logical options.  Based on Andy's list, I propose the following (only minor differences):
-s, --server server
        URL of the WBEM server.
        - Scheme is optional and defaults to HTTP.
        - Host is required.
        - Port is optional and defaults to the standard ports, depending on HTTP/HTTPS.
**OR it may be an positional parameter. Since this is the core piece of information for a connection
   and the only alternates would be to defualt to something like localhost it would seem that making this positional (depending the the app is logical. Some of the reasons for having the arg as option would be, for example, in the case where the goal was to communicate with multiple servers.  Note that I have not thought out the question of merging this with a config file so that may be a reason to allow or force the server optional arg.**
-n, --namespace namespacekeing
        Namespace in the WBEM server to work against
        Default: root/cimv2
-u, --username username
        Username for authenticating with the WBEM server
        Default: No username
-p, --password password
        Password for authenticating with the WBEM server
        Default: Prompt for password if a username was specified
-c, --certfile certfile
        File path of certificate file to be used with HTTPS.
**OR may be only the --certfile, i.e. no -c option since this is not used often**
-k, --keyfile keyfile
        File path of key file to be used with HTTPS.
**OR may be only the --certfile, i.e. no -k option since this is not used often**
-t, --timeout timeout
        Timeout in seconds
**Add**
--ca_certs
    File path of certificate file or directory used to validate WBEM Server
-no --no_verification
   Boolean that specifies that the client is to ignore any certificates received from the server
        Default: TBD

**Comment by andy-maier** [comment]
I agree that `server` should be positional.
See my other comments on PR #279.

**Comment by andy-maier** [comment]
At this point, the `wbemcli` options as changed in PR #279 are the master definition of where we want to go.

**Comment by andy-maier** [comment]
PR #279 fixes this for the wbemcli command.

The mof_compiler command still needs to be fixed.

**Comment by KSchopmeyer** [comment]
We still have the mof compiler and we may also have some sort of cimcli cmdline in the future. Keep this one open

**Comment by andy-maier** [comment]
Here is my proposal on how to change the command line interface for `mof_compiler`:

1. Add the following new options in the same way they exist on `wbemcli`:
   * `-nvc, --no-verify-cert`
   * `--cacerts cacerts`
   * `--certfile certfile`
   * `--keyfile keyfile`

2. Change the following existing options:
   * `-u url, --url url`: Rename to `-s url, --server url` (the server is a positional argument in `wbemcli`, but in `mof_compiler`, the positional arguments are used for the MOF files). Also, it would no longer have a default, i.e. it is a required option.
   * `-l username, --username username`: Rename to `-u user, --user user` (to become consistent with `wbemcli`)
   * `-s dir, --search dir`: Rename to `-I dir, --include dir` (to use something meaningful, given that `-s` is now used for the server URL.

3. Other changes:
   * Default for password will be to be prompted for, if a user was specified. This is consistent with `wbemcli`.
   * namespace will now have a default of `"root/cimv2"` instead of being required. This is consistent with `wbemcli`.

This results in the following overall command line help for `mof_compiler`:

```
usage: mof_compiler [options] moffile ...

Compile MOF files, and update the repository of a WBEM server with the result.

Positional arguments:
  moffile               Path name of the MOF file to be compiled.
                        Can be specified multiple times.

Server related options:
  Specify the WBEM server and namespace, and access related information.

  -s url, --server url  Required: Host name or url of the WBEM server in this format:
                            [{scheme}://]{host}[:{port}]
                        - scheme: Defines the protocol to use;
                            - "https" for HTTPs protocol
                            - "http" for HTTP protocol.
                          Default: "https".
                        - host: Defines host name as follows:
                             - short or fully qualified DNS hostname,
                             - literal IPV4 address(dotted)
                             - literal IPV6 address (RFC 3986) with zone
                               identifier extensions(RFC 6874)
                               supporting "-" or %25 for the delimiter.
                        - port: Defines the WBEM server port to be used
                          Defaults:
                             - HTTP  - 5988
                             - HTTPS - 5989
  -n namespace, --namespace namespace
                        Namespace in the WBEM server to work against
                        Default: root/cimv2
  -u user, --user user  User name for authenticating with the WBEM server.
                        Default: No user name.
  -p password, --password password
                        Password for authenticating with the WBEM server.
                        Default: Will be prompted for, if user name
                        specified.
  -nvc, --no-verify-cert
                        Client will not verify certificate returned by the
                        WBEM server (see cacerts). This bypasses the client-
                        side verification of the server identity, but allows
                        encrypted communication with a server for which the
                        client does not have certificates.
  --cacerts cacerts     File or directory containing certificates that will be
                        matched against a certificate received from the WBEM
                        server. Set the --no-verify-cert option to bypass
                        client verification of the WBEM server certificate.
                        Default: Searches for matching certificates in the
                        following system directories:
                        /etc/pki/ca-trust/extracted/openssl/ca-bundle.trust.crt
                        /etc/ssl/certs
                        /etc/ssl/certificates
  --certfile certfile   Client certificate file for authenticating with the
                        WBEM server. If option specified the client attempts
                        to execute mutual authentication.
                        Default: Simple authentication.
  --keyfile keyfile     Client private key file for authenticating with the
                        WBEM server. Not required if private key is part of the
                        certfile option. Not allowed if no certfile option.
                        Default: No client key file. Client private key should
                        then be part  of the certfile
  -t timeout, --timeout timeout
                        Timeout of the completion of WBEM Server operation
                        in seconds(integer between 0 and 300).
                        Default: No timeout

Action related options:
  Specify actions against the repository. Default: create/update elements.

  -r, --remove          Remove elements (found in the MOF files) from the
                        repository, instead of creating or updating them
  -d, --dry-run         Don't actually modify the repository, just check MOF
                        syntax. Connection to WBEM server is still required to
                        check qualifiers.

General options:
  -I dir, --include dir Path name of an additional include directory for MOF
                        include files. Can be specified multiple times.
  -v, --verbose         Print more messages while processing
  -h, --help            Show this help message and exit

Example: mof_compiler CIM_Schema_2.45.mof -s https://localhost:15989 -n
root/cimv2 -u sheldon -p penny42
```

**Comment by andy-maier** [comment]
Please comment on the proposal above, before I start coding it.

**Comment by andy-maier** [comment]
Karl wants to review the proposal and provide feedback.

**Comment by andy-maier** [comment]
We agreed in today's call to go ahead with the proposal.

**Comment by andy-maier** [comment]
Fixed in PR #385.
